likelihood/cl_like/bacco.py

```diff
@@ -0,0 +1,335 @@
+
+    Args:
+        log10k_min (float): decimal logarithm of the minimum
+            Fourier scale (in Mpc^-1) for which you want to
+            calculate perturbation theory quantities.
+        log10k_max (float): decimal logarithm of the maximum
+            Fourier scale (in Mpc^-1) for which you want to
+            calculate perturbation theory quantities.
+    """
+    def __init__(self, log10k_min=-2, log10k_max=0, nk_per_decade=20, h=None, k_filter=None):
+
+        self.h = h
+        if np.log10(10**log10k_min/self.h) < -2:
+            logger.info('Setting k_min to BACCO default.')
+            log10k_min = np.log10((1e-2)*self.h)
+        if np.log10(10**log10k_max/self.h) > np.log10(0.75):
+            logger.info('Setting k_max to BACCO default.')
+            log10k_max = np.log10(0.75*self.h)
+        nk_total = int((log10k_max - log10k_min) * nk_per_decade)
+        self.ks = np.logspace(log10k_min, log10k_max, nk_total)
+        self.bacco_table = None
+
+        # redshifts for creating the Pk-2d object
+        z = np.array([1.5, 1.0, 0.85, 0.7, 0.55, 0.4, 0.25, 0.1, 0.0])
+        self.a_s = 1. / (1. + z)
+        if k_filter is not None:
+            self.wk_low = 1-np.exp(-(self.ks/k_filter)**2)
+        else:
+            self.wk_low = np.ones(nk_total)
+
+
+    def _cosmo_to_bacco(self, cosmo):
+        pars = {
+            'omega_matter': cosmo['Omega_b']+cosmo['Omega_c'], # No massive neutrinos
+            'omega_baryon': cosmo['Omega_b'],
+            'hubble': cosmo['H0']/100.,
+            'ns': cosmo['n_s'],
+            #TODO: Remove massive neutrinos from sigma_8
+            'sigma8': cosmo['sigma8'],
+            'neutrino_mass': np.sum(cosmo['m_nu']),
+            'w0': cosmo['w0'],
+            'wa': cosmo['wa'],
+            'expfactor': 1.  # random scale factor just to initialize
+        }
+        return pars
+
+    def update_pk(self, cosmo):
+        """ Update the internal PT arrays.
+
+        Args:
+            pk (array_like): linear power spectrum sampled at the
+                internal `k` values used by this calculator.
+        """
+
+        # translate the pyccl cosmology parameters into bacco notation
+        pars = self._cosmo_to_bacco(cosmo)
+        #TODO: Do this only once?
+        lpt_emu = baccoemu.Lbias_expansion()
+        # convert k_s [Mpc^-1] into h Mpc^-1 units just for the calculation
+        k = self.ks / pars['hubble']
+
+        # 10 redshifts, 15 combinations between bias params, and the ks (added b0)
+        num_comb = 15
+        pk2d_bacco = np.zeros((len(self.a_s), num_comb, len(self.ks)))
+
+        # compute the power for each redshift
+        # t1 = time.time()
+        for i in range(len(self.a_s)):
+            pars['expfactor'] = self.a_s[i]
+            # call the emulator of the nonlinear 15 lagrangian bias expansion terms, shape is (15, len(k))
+            k, pnn = lpt_emu.get_nonlinear_pnn(pars, k=k)
```

Comment:
Is the output k the same as the input k?

Reply:
Also, is this call internally calling CLASS/CAMB or something like that? I thought BACCO basically uses boost factors applied onto the linear or non-linear power spectra, so if we're not supplying these, does it need to call CAMB?

Reply:
Yes this is the same k. I've changed this line in the latest commit to `_` since we don't need it.

Reply:
Regarding CLASS/CAMB: no, I think the way this works is that the HEFT emulator indeed only emulates the ratios, but the call `get_nonlinear_pnn` also calls the pure LPT emulator and then combines these two into the HEFT predictions.